File: Chapter12/VB/HttpService/Teams/Connection/RequestQueue.vb
Imports System.Collections
Imports System.Threading

' ThreadSemaphore is used by RequestQueue, and encapsulates
' the items required to managed blocked request threads

Public Class ThreadSemaphore
  Public requestEvent As New ManualResetEvent(False)
  Public sequenceNumber As Integer
  Public request As String
  Public response As Byte()
  Public inUse As Boolean
  Public startTime As DateTime

  Public Function Wait(ByVal theDuration As TimeSpan) As Boolean
    Return requestEvent.WaitOne(theDuration, True)
  End Function

  Public Sub Signal()
    requestEvent.Set()
  End Sub
End Class

' RequestQueue maintains a list of blocked request threads.
' When a client sends a request, the thread is blocked and 
' a ThreadSemaphore object is kept in a hashtable. When the
' response returns, the ThreadSemaphore is signaled, allowing
' the request thread to continue and return data to the 
' calling client.

Public Class RequestQueue

  Private Shared _singleton As RequestQueue
  Public Shared ReadOnly Property Singleton() As RequestQueue
    Get
      If _singleton Is Nothing Then
        _singleton = New RequestQueue
      End If
      Return _singleton
    End Get
  End Property

  ' a pool of prebuilt semaphores used with blocked requests.
  ' We support up to 500 concurrent requests. The value is arbitrary
  Private _semaphores(500) As ThreadSemaphore

  ' key is SequenceNumber, value is RequestSemaphore
  Private _blockedRequests As New Hashtable

  Public Sub New()
    For i As Integer = 0 To _semaphores.Length - 1
      Dim semaphore As New ThreadSemaphore
      semaphore.requestEvent = New ManualResetEvent(False)
      _semaphores(i) = semaphore
    Next
  End Sub

  Public Function Add(ByVal theSequenceNumber As Integer, ByVal theRequest As String) As ThreadSemaphore
    SyncLock (Me)
      Dim semaphore As ThreadSemaphore = GetFirstAvailableSemaphore()
      _blockedRequests.Add(theSequenceNumber, semaphore)
      semaphore.requestEvent.Reset()
      semaphore.sequenceNumber = theSequenceNumber
      semaphore.request = theRequest
      semaphore.response = Nothing
      semaphore.startTime = DateTime.Now
      Return semaphore
    End SyncLock
  End Function

  Function GetFirstAvailableSemaphore() As ThreadSemaphore
    SyncLock (Me)
      For Each semaphore As ThreadSemaphore In _semaphores
        If Not semaphore.inUse Then
          semaphore.inUse = True
          Return semaphore
        End If
      Next
    End SyncLock
    Throw New Exception("RequestQueue: No semaphores available")
  End Function

  Public Sub Remove(ByVal theSequenceNumber As Integer)
    SyncLock (Me)
      Dim semaphore As ThreadSemaphore = DirectCast(_blockedRequests(theSequenceNumber), ThreadSemaphore)
      _blockedRequests.Remove(theSequenceNumber)
      semaphore.inUse = False
    End SyncLock
  End Sub

  Public Function [Get](ByVal theSequenceNumber As Integer) As ThreadSemaphore
    Return DirectCast(_blockedRequests(theSequenceNumber), ThreadSemaphore)
  End Function
End Class

File: Chapter12/C#/HttpService/Teams/Connection/RequestQueue.cs
using System;
using System.Collections;
using System.Threading;

namespace HttpService
{
  /// <summary>
  /// ThreadSemaphore is used by RequestQueue, and encapsulates
  /// the items required to managed blocked request threads
  /// </summary>
  public class ThreadSemaphore
  {
    public ManualResetEvent requestEvent = new ManualResetEvent(false);
    public int sequenceNumber;
    public string request;
    public byte[] response;
    public bool inUse;
    public DateTime startTime;

    public bool Wait(TimeSpan theDuration)
    {
      return requestEvent.WaitOne(theDuration, true);
    }

    public void Signal()
    {
      requestEvent.Set();
    }
  }

  /// <summary>
  /// RequestQueue maintains a list of blocked request threads.
  /// When a client sends a request, the thread is blocked and 
  /// a ThreadSemaphore object is kept in a hashtable. When the
  /// response returns, the ThreadSemaphore is signaled, allowing
  /// the request thread to continue and return data to the 
  /// calling client.
  /// </summary>
  public class RequestQueue
  {
    static RequestQueue singleton;
    public static RequestQueue Singleton
    {
      get 
      {
        if (singleton == null)
          singleton = new RequestQueue();
        return singleton;
      }
    }

    // a pool of prebuilt semaphores used with blocked requests
    ThreadSemaphore[] semaphores;

    // key is SequenceNumber, value is RequestSemaphore
    Hashtable blockedRequests = new Hashtable();

    public RequestQueue()
    {
      // we support up to 500 concurrent requests. The value is arbitrary
      semaphores = new ThreadSemaphore[500];  
      for (int i = 0; i < semaphores.Length; i++)
      {
        ThreadSemaphore semaphore = new ThreadSemaphore();
        semaphore.requestEvent = new ManualResetEvent(false);
        semaphores[i] = semaphore;
      }
    }
    
    public ThreadSemaphore Add(int theSequenceNumber, string theRequest)
    {
      lock(this)
      {
        ThreadSemaphore semaphore = GetFirstAvailableSemaphore();
        blockedRequests.Add(theSequenceNumber, semaphore);
        semaphore.requestEvent.Reset();
        semaphore.sequenceNumber = theSequenceNumber;
        semaphore.request = theRequest;
        semaphore.response = null;
        semaphore.startTime = DateTime.Now;
        return semaphore;
      }
    }

    ThreadSemaphore GetFirstAvailableSemaphore()
    {
      lock(this)
      {
        foreach (ThreadSemaphore semaphore in semaphores)
        {
          if (semaphore.inUse) continue;
          semaphore.inUse = true;
          return semaphore;
        }
      }
      throw new Exception("RequestQueue: No semaphores available");
    }

    public void Remove(int theSequenceNumber)
    {
      lock(this)
      {
        ThreadSemaphore semaphore = blockedRequests[theSequenceNumber] as ThreadSemaphore;
        blockedRequests.Remove(theSequenceNumber);
        semaphore.inUse = false;
      }
    }

    public ThreadSemaphore Get(int theSequenceNumber)
    {
      return blockedRequests[theSequenceNumber] as ThreadSemaphore;
    }
  }
}
